File: feel/feelmesh/hypercube.cpp
/* -*- mode: c++; coding: utf-8; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; show-trailing-whitespace: t -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4 

  This file is part of the Feel library

  Author(s): Christophe Prud'homme <christophe.prudhomme@ujf-grenoble.fr>
       Date: 2006-02-20

  Copyright (C) 2006 EPFL

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3.0 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
*/
/**
   \file hypercube.cpp
   \author Christophe Prud'homme <christophe.prudhomme@ujf-grenoble.fr>
   \date 2006-02-20
 */
#include <feel/feelmesh/hypercube.hpp>

namespace Feel
{

/// \cond DETAIL
namespace details
{

/**
 * \class quad

\code
          3   2   2
           |-----|
           |     |
         3 |     | 1
           |     |
         0 ------- 1
              0
\endcode
*/
template<uint16_type Order >
const uint16_type quad<Order>::__e2p_order1[8] =
    {
        0, 1, // points in edge 0
        1, 2, // points in edge 1
        2, 3, // points in edge 2
        3, 0  // points in edge 3
    };

template<uint16_type Order >
const uint16_type quad<Order>::__e2p_order2[12] =
    {
        0, 1, 4, // points in edge 0
        1, 2, 5, // points in edge 1
        2, 3, 6, // points in edge 2
        3, 0, 7  // points in edge 3
    };

template<uint16_type Order >
const uint16_type quad<Order>::__e2p_order3[16] =
    {
        0, 1,  4,  5, // points in edge 0
        1, 2,  6,  7, // points in edge 1
        2, 3,  8,  9, // points in edge 2
        3, 0, 10, 11  // points in edge 3
    };


template<uint16_type Order >
const uint16_type quad<Order>::__f2p[4] =
    {
        0, // point 0
        1, // point 1
        2, // point 2
        3  // point 3
    };
template<uint16_type Order >
const uint16_type quad<Order>::__f2e[4] =
    {
        0, // edge 0
        1, // edge 1
        2, // edge 2
        3  // edge 3
    };


/**
   \class hexa

\code
        7-------6
       /.      /|
      / .     / |
     4_______5  |
     |  .    |  |
     |  3....|..2
     | .     | /
     |.      |/
     0_______1
\endcode
*/


// edge to point relation
template<uint16_type Order >
const uint16_type hexa<Order>::__e2p_order1[24] =
    {
        0, 1,    // edge 0
        1, 2,    // edge 1
        2, 3,    // edge 2
        3, 0,    // edge 3
        1, 5,    // edge 4
        5, 4,    // edge 5
        4, 0,    // edge 6
        2, 6,    // edge 7
        6, 5,    // edge 8
        3, 7,    // edge 9
        7, 6,    // edge 10
        4, 7     // edge 11
    };

template<uint16_type Order >
const uint16_type hexa<Order>::__e2p_order2[36] =
    {
        0, 1, 8,    // edge 0
        1, 2, 9,    // edge 1
        2, 3, 10,   // edge 2
        3, 0, 11,   // edge 3
        1, 5, 12,   // edge 4
        5, 4, 13,   // edge 5
        4, 0, 14,   // edge 6
        2, 6, 15,   // edge 7
        6, 5, 16,   // edge 8
        3, 7, 17,   // edge 9
        7, 6, 18,   // edge 10
        4, 7, 19    // edge 11
    };


template<uint16_type Order >
const uint16_type hexa<Order>::__e2p_order3[48] =
    {
        0, 1, 8, 9,     // edge 0
        1, 2, 10, 11,   // edge 1
        2, 3, 12, 13,   // edge 2
        3, 0, 14, 15,   // edge 3
        1, 5, 16, 17,   // edge 4
        5, 4, 18, 19,   // edge 5
        4, 0, 20, 21,   // edge 6
        2, 6, 22, 23,   // edge 7
        6, 5, 24, 25,   // edge 8
        3, 7, 26, 27,   // edge 9
        7, 6, 28, 29,   // edge 10
        4, 7, 30, 31    // edge 11
    };


// face to point relation
template<uint16_type Order >
const uint16_type hexa<Order>::__f2p_order1[24] =
    {
        0, 1, 2, 3, // face 0
        0, 1, 5, 4, // face 1
        1, 2, 6, 5, // face 2
        2, 3, 7, 6, // face 3
        3, 0, 4, 7, // face 4
        4, 5, 6, 7  // face 5
    };

template<uint16_type Order >
const uint16_type hexa<Order>::__f2p_order2[54] =
    {
        0, 1, 2, 3,  8, 9,  10, 11, 20, // face 0
        0, 1, 5, 4,  8, 12, 13, 14, 21, // face 1
        1, 2, 6, 5,  9, 15, 16, 12, 22, // face 2
        2, 3, 7, 6, 10, 17, 18, 15, 23, // face 3
        3, 0, 4, 7, 11, 14, 19, 17, 24, // face 4
        4, 5, 6, 7, 13, 16, 18, 19, 25  // face 5
    };


template<uint16_type Order >
const uint16_type hexa<Order>::__f2p_order3[96] =
    {
        0, 1, 2, 3,  8,  9, 10, 11, 12, 13, 14, 15, 32, 33, 34, 35, // face 0
        0, 1, 5, 4,  8,  9, 16, 17, 18, 19, 20, 21, 36, 37, 38, 39, // face 1
        1, 2, 6, 5, 10, 11, 22, 23, 24, 25, 17, 16, 40, 41, 42, 43, // face 2
        2, 3, 7, 6, 12, 13, 26, 27, 28, 29, 23, 22, 44, 45, 46, 47, // face 3
        3, 0, 4, 7, 14, 15, 21, 20, 30, 31, 27, 26, 48, 49, 50, 51, // face 4
        4, 5, 6, 7, 19, 18, 25, 24, 29, 28, 31, 30, 52, 53, 54, 55  // face 5
    };

// face to edge relation
template<uint16_type Order >
const uint16_type hexa<Order>::__f2e[24] =
    {
        0,  1,  2,  3, // face 0
        0,  4,  5,  6, // face 1
        1,  7,  8,  4, // face 2
        2,  9, 10,  7, // face 3
        3,  6, 11,  9, // face 4
        5,  8, 10, 11  // face 5
    };

// edge permutation in face
template<uint16_type Order >
const int16_type hexa<Order>::__f2e_permutation[24] =
    {
        1,  1,  1,  1, // face 0
        1,  1,  1,  1, // face 1
        1,  1,  1, -1, // face 2
        1, -1,  1, -1, // face 3
        1, -1,  1, -1, // face 4
       -1, -1, -1, -1  // face 5
    };

} // details
/// \endcond

} // Feel

template class Feel::details::quad<1>;
template class Feel::details::quad<2>;
template class Feel::details::quad<3>;
template class Feel::details::quad<4>;
template class Feel::details::quad<5>;

template class Feel::details::hexa<1>;
template class Feel::details::hexa<2>;
template class Feel::details::hexa<3>;
template class Feel::details::hexa<4>;
template class Feel::details::hexa<5>;

File: feelpp/tools/polyvis/hypercube.py
#! /usr/bin/python

# -*- mode: python -*-
#
#  This file is part of the Feel library
#
#  Author(s): Goncalo Pena <gpena@mat.uc.pt>
#        Date: 2010-01-09
#
#   Copyright (C) 2010 University of Coimbra
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Lesser General Public
#   License as published by the Free Software Foundation; either
#   version 2.1 of the License, or (at your option) any later version.
#
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public
#   License along with this library; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
#   University of Coimbra
#
# \file hypercube.py
# \author Goncalo Pena <gpena@mat.uc.pt>
# \date 2010-01-09
#

# This file contains the functions to draw the contours of a hypercube,
# draw the equispaced pointset up to any order,
# and draw the numbering of the points (in the pointset) by subentity

from pyx import *
import math
from scipy import *


def line(c, x0, x1):
    a, b = x0
    d, e = x1
    result = []

    sqa = path.path( path.moveto(a, b), path.lineto(d,e), path.closepath())

    c.stroke(sqa, [style.linewidth(0.04)])

    return result


def border(c, x0, line_size):
    a, b = x0
    result = []

    #border of square

    sqa = path.path( path.moveto(a, b), path.lineto(a+line_size, b),
                     path.lineto(a+line_size, b+line_size), path.lineto(a, b+line_size),
                     path.closepath())

    c.stroke(sqa, [style.linewidth(0.04)])

    return result


def borderRect(c, x0, line_size1, line_size2):
    a, b = x0
    result = []

    #border of square

    sqa = path.path( path.moveto(a, b), path.lineto(a+line_size1, b),
                     path.lineto(a+line_size1, b+line_size2), path.lineto(a, b+line_size2),
                     path.closepath())

    c.stroke(sqa, [style.linewidth(0.04)])

    return result


def Vertex(c, line_size, origin, local_id, circle_radius):

    result = []

    x = origin[0]
    y = origin[1]

    if local_id == 1:
	x = x + line_size
    else:
	if local_id == 2:
	    x = x + line_size
	    y = y + line_size
	else:
	    if local_id == 3:
	        y = y + line_size

    c.stroke(path.circle(x, y, circle_radius), [style.linewidth.thin, deco.filled([color.rgb.red])])

    return result


def Edge(c, line_size, origin, local_id, order, circle_radius):
    # plots the points in an edge of a triangle


    result = []

    h = 1.0*line_size/order

    x0 = origin[0]
    x1 = origin[1]

    if local_id == 0:
	v = [1,0]
    else:
	if local_id == 1:
	    x0 = x0 + line_size
    	    v=[0,1]
        else:
	    if local_id == 2:
		x0 = x0 + line_size
		x1 = x1 + line_size
	        v=[-1,0]
	    else:
		x1 = x1 + line_size
	        v=[0,-1]


    for i in range(order-1):
	p0 = x0 + (i+1)*h*v[0]
	p1 = x1 + (i+1)*h*v[1]

	c.stroke(path.circle(p0, p1, circle_radius), [style.linewidth.thin, deco.filled([color.rgb.blue])])


    return result


def Face(c, line_size, origin, order, circle_radius):

    result = []

    if order >= 2:
    	h = 1.0*line_size/(order)

    	for i in range(order-1):
           for j in range(order-1):
	       p0 = origin[0] + 1.0*h*( (i+1)  )
	       p1 = origin[1] + 1.0*h*( (j+1) )

	       c.stroke(path.circle(p0, p1, circle_radius), [style.linewidth.thin, deco.filled([color.rgb.green])])

    return result


def numberLine(c, line_size, origin, order, circle_radius, shift):

    result = []

    local_id = 0

    x = origin[0]
    y = origin[1]

    h = 1.0*line_size/order

    x = origin[0] - shift
    y = origin[1] - 4.0*shift
    v = [1,0]
    first = 2

    for i in range(order-1):
        p0 = x + (i+1)*h*v[0]
        p1 = y + (i+1)*h*v[1]

        c.text(p0, p1, i+first, [text.halign.boxleft])

    return result

def numberVertex(c, line_size, origin, local_id, circle_radius, shift):

    result = []

    x = origin[0] - shift
    y = origin[1] - 3.0*shift

    if local_id == 0:
	c.text(x, y, local_id, [text.halign.boxright])

    if local_id == 1:
	x = x + line_size + 2.0*shift
        c.text(x, y, local_id, [text.halign.boxleft])

    if local_id == 2:
	x = x + line_size + 4.0*shift
	y = y + line_size + 4.0*shift
        c.text(x, y, local_id, [text.halign.boxright])

    if local_id == 3:
	y = y + line_size + 4.0*shift
        c.text(x, y, local_id, [text.halign.boxright])

    return result

def numberEdge(c, line_size, origin, local_id, order, circle_radius, shift):

    result = []

    x = origin[0]
    y = origin[1]

    h = 1.0*line_size/order

    if local_id == 0:
	x = x - shift
	y = y - 4.0*shift
	v = [1,0]
	first = 4
    else:
	if local_id == 1:
	    x = x + line_size + 2.0*shift
	    y = y - shift
    	    v=[0,1]
	    first = order + 3
        else:
	    if local_id == 2:
		x = x + line_size
		y = y + line_size + 2.0*shift
	        v=[-1,0]
		first = 2*order +2
	    else:
		x = x - 2.0*shift
		y = y + line_size - shift
	        v=[0,-1]
		first = 3*order +1


    for i in range(order-1):
	p0 = x + (i+1)*h*v[0]
	p1 = y + (i+1)*h*v[1]

	if local_id == 0:
	    c.text(p0, p1, i+first, [text.halign.boxleft])
	else:
	    if local_id == 1:
		c.text(p0, p1, i+first, [text.halign.boxleft])
	    else:
		if local_id == 2:
		    c.text(p0, p1, i+first, [text.halign.boxcenter])
		else:
		    c.text(p0, p1, i+first, [text.halign.boxright])


    return result

def numberFace(c, line_size, origin, order, circle_radius, shift):

    result = []

    x = origin[0]
    y = origin[1]

    if order >= 2:
    	h = 1.0*line_size/(order)


        p = 4 + 4*(order-1)

        for i in range(order-1):
            for j in range(order-1):
	        p0 = origin[0] + 1.0*h*( i+1 ) - 2.0*shift
	        p1 = origin[1] + 1.0*h*( j+1 ) - 1.0*shift

	        c.text(p0, p1, p, [text.halign.boxright])
	        p = p+1


    return result


def n_side_points(order):
    return order-1

def identity(order, i):
    return i

def base(order, index):
    N = n_side_points(order)

    p = 0

    result = index

    for i in range(N):
	k = N-1-i

	first = k * N

	for j in range(N):
	    if index == p:
		result = first+j

   	    p = p+1

    return result


def once_anticlockwise(order, index):
    N = n_side_points(order)

    p = 0

    for i in range(N):
	k = N-1-i

	for j in range(N):
	    if index == p:
		return k + N*j

   	    p = p+1


def second_diagonal(order, index):
    p = base(order, index)
    return once_anticlockwise(order, p)


def height(order, index):
    p = second_diagonal(order, index)
    return once_anticlockwise(order, p)

def twice_clockwise(order, index):
    p = base(order, index)
    return height(order, p)

def diagonal(order, index):
    p = height(order, index)
    return once_anticlockwise(order, p)

def once_clockwise(order, index):
    p = once_anticlockwise(order, index)
    return twice_clockwise(order, p)

def addIndices(c, index, line_size, origin, order, circle_radius, shift):

    result = []

    x = origin[0]
    y = origin[1]

    if order >= 2:
    	h = 1.0*line_size/(order)

    p = 0
    index_shift = 0 #4 + 4*(order-1)


    for i in range(order-1):
        for j in range(order-1):
	    p0 = origin[0] + 1.0*h*( i+1 ) - 2.0*shift
	    p1 = origin[1] + 1.0*h*( j+1 ) - 1.0*shift

	    if index == 0:
	    	c.text(p0, p1, index_shift + identity(order, p), [text.halign.boxright])

	    if index == 1:
	    	c.text(p0, p1, index_shift + second_diagonal(order, p), [text.halign.boxright])

	    if index == 2:
	    	c.text(p0, p1, index_shift + base(order, p), [text.halign.boxright])

	    if index == 3:
	    	c.text(p0, p1, index_shift + once_anticlockwise(order, p), [text.halign.boxright])

	    if index == 4:
	    	c.text(p0, p1, index_shift + diagonal(order, p), [text.halign.boxright])

	    if index == 5:
	    	c.text(p0, p1, index_shift + twice_clockwise(order, p), [text.halign.boxright])
	    if index == 6:
	    	c.text(p0, p1, index_shift + once_clockwise(order, p), [text.halign.boxright])
	    if index == 7:
	    	c.text(p0, p1, index_shift + height(order, p), [text.halign.boxright])
	    p = p+1
    return result

















def addLegend(c, index, x0, hpos, vpos):
    result = []

    p0 = x0[0] + hpos
    p1 = x0[1] - vpos

    if index == 0:
    	c.text(p0, p1, "$\omega_0$", [text.halign.boxcenter])
    if index == 1:
    	c.text(p0, p1, "$\omega_1$", [text.halign.boxcenter])
    if index == 2:
    	c.text(p0, p1, "$\omega_2$", [text.halign.boxcenter])
    if index == 3:
    	c.text(p0, p1, "$\omega_3$", [text.halign.boxcenter])
    if index == 4:
    	c.text(p0, p1, "$\omega_4$", [text.halign.boxcenter])
    if index == 5:
    	c.text(p0, p1, "$\omega_5$", [text.halign.boxcenter])
    if index == 6:
    	c.text(p0, p1, "$\omega_6$", [text.halign.boxcenter])
    if index == 7:
    	c.text(p0, p1, "$\omega_7$", [text.halign.boxcenter])


    return result

# Local Variables:
# indent-tabs-mode: t
# End:

